### ISSUE TITLE
regression model predicts all samples the same value

### ISSUE THREAD

**Issue by bosmont** [topepo/caret#99]
Hi,

I am following up with a question I posted previously (#66).

I built a regression model as follow:

Control <- trainControl(method = "repeatedcv",repeats=10,number=5, verboseIter =TRUE)
netFit <- train(x = expr.train, y = TC.train,
                  method = "glmnet", 
                  preProcess=c('center', 'scale'),
                  type.gaussian="naive", family="gaussian",
                  tuneGrid = enetGrid,
                  trControl = Control)

the training data contain 30 samples with ~3550 features for each sample.  
when the command finished, I got following error:

Aggregating results
Selecting tuning parameters
Fitting alpha = 1e-08, lambda = 2981 on full training set
Warning message:
In nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo,  :
  There were missing values in resampled performance measures.

I am not sure about the fitting alpha and lambda values here since they are the smallest alpha and the largest lambda I provided in the parameter grid.  And as Max pointed out in my previous question, I do see all samples to be predicted the same value (the average of values in training samples), and all values in the Rsquared column in the resample sub-object are "NA"s.  I do not understand why this is happening, and how do I fix it?

Here is my Rsession:

Here is my R session:
R version 3.1.2 (2014-10-31)
Platform: x86_64-unknown-linux-gnu (64-bit)

locale:
[1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C LC_TIME=en_US.UTF-8 LC_COLLATE=en_US.UTF-8 LC_MONETARY=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8

[7] LC_PAPER=en_US.UTF-8 LC_NAME=C LC_ADDRESS=C LC_TELEPHONE=C LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] parallel stats graphics grDevices utils datasets methods base

other attached packages:
[1] gdata_2.13.3 reshape2_1.4 glmnet_1.9-8 Matrix_1.1-4 car_2.0-22 doMC_1.3.3 iterators_1.0.7 foreach_1.4.2 caret_6.0-37 ggplot2_1.0.0

[11] lattice_0.20-29

loaded via a namespace (and not attached):
[1] BradleyTerry2_1.0-5 brglm_0.5-9 codetools_0.2-9 colorspace_1.2-4 compiler_3.1.2 digest_0.6.4 grid_3.1.2 gtable_0.1.2

[9] gtools_3.4.1 labeling_0.3 lme4_1.1-7 MASS_7.3-35 minqa_1.2.4 munsell_0.4.2 nlme_3.1-118 nloptr_1.0.4

Thanks so much for your help.

**Comment by zachmayer** [comment]
Without sample code to reproduce the error, I am just speculating, but `glmnet` probably dropped all the terms out of your model and fit an intercept-only model.  This would result in constant-valued predictions,  which would result in NA R2 values (as a prediction with zero variance has undefined R2).

**Comment by zachmayer** [comment]
Also, please do not open new issues.  Update existing issues.

**Comment by bosmont** [comment]
Thanks for your quick response.  Could you let me know why glmnt dropped all the predictors?  I checked the beta object in the finalModel, and there are a lot of nonzero values.  

Besides, I do not understand why the beta object is a matrix with 100 columns (and n number of rows, with n as the number of predictors, that part I can see).  Where is the 100 from?

Thanks again.

**Comment by topepo** [comment]
glmnet saves a lot of different sets of parameter estimates and the exact
ones aren't needed until the time of prediction.

The model may have preferred a model with only intercept if it felt that
there were no informative predictors. For example, trees can do this too;
if there is no viable split, no is returned.

**Comment by bosmont** [comment]
So out of the 100 sets of beta values that are saved in finalModel$beta object, how do I know which one is used to predict the new samples?

**Comment by bosmont** [comment]
This would be important if I want to do feature selection based on the beta values.

**Comment by topepo** [comment]
The `bestTune` sub-object has the optimal settings and the use the `coef`
method to get the corresponding coefficients.

You might want to take a look at this book:

   http://www-bcf.usc.edu/~gareth/ISL/

The book is free and the authors focus on this particular model. This will
probably help clear up how this model function works. A few of your
questions are really about `glmnet` and not `train`.

**Comment by bosmont** [comment]
Thanks for your advice.  will read the book.

In the meantime, would appreciate it if you could answer another question here.  In the "bestTune" object, there are optimized values of both alpha, and lamdba.  And I got different sets of beta when running:

coef(netFit2$finalModel, netFit2$bestTune$alpha) 
and 
coef(netFit2$finalModel, netFit2$bestTune$lambda)

How do I get the coefficients that are coresponding to both best alpha and best lambda?

**Comment by zachmayer** [comment]
Read ?coef.glment

**Comment by zachmayer** [comment]
To follow up: this is not the place to ask question about `glmnet`.  The best forum for those questions is probably [cross-validated](http://stats.stackexchange.com/) or [stackoverflow](http://stackoverflow.com/questions/tagged/r). However, `glmnet` happens to be one of my favorite modeling packages, so I'll educate you a little. 

`glmnet` is cool because:
1. It can do both Lasso regularization (L1, feature selection) and ridge regularization (L2, or shrinkage), and any combination of the 2.  The type of regularization is controlled by the alpha penalty, and, for a given `glmnet` model, is fixed.
2. It can fit the ENTIRE regularization path, from the full model (as would be returned by `glm`) to the intercept-only model (as would be fit by `mean`).  The amount, or strength of the regularization is controlled by the lambda parameter, and the final `glmnet` model can predict for any lambda value.

So, the final `glmnet` model gives you a hundred or so models for the the price of one, which is very, very cool.  Of course, when you go to make predictions or choose coefficients, you much provide a lambda value to the final model.  One way to choose a lambda is through cross-validation, using `cv.glmnet` or `caret::train`.  The latter has a small advantage in that it also allows you to optimize the `alpha` parameter.
   
So, to re-iterate: the final `glmnet` object has a fixed alpha and variable lambda.  Both `predict.glmnet` and `coef.glment` require a lambda parameter to decide which model to use for predictions or coefficients.
    
ALL of this information is provided in `?glmnet`, `?cv.glmnet`, `?predict.glmnet` and `?predict. cv.glmnet`.  `?plot.glmnet` is cool too. You have to work a little to extract the information you need from those help pages, but that's a valuable skill to develop.
